library/Core/src/Juvix/Core/Common/Context.hs

```diff
@@ -301,137 +301,120 @@ mapWithContextPure ::
+-- records are treated as if they are part of the module above it if
+-- there is a signature!  See @mapCurrentContext@ for more information
+-- about specifics, as we just dispatch to that for the real changes
+mapWithContext ::
+  Monad m => T term ty sumRep -> ContextForms m term ty sumRep -> m (T term ty sumRep)
+mapWithContext t f =
+  -- If we map over every current module in the top level map, we map
+  -- across the entire context!
+  overTopLevelMap t (mapCurrentContext f)
+
+-- | @mapWithName@ maps through the context definitions, supplying the
+-- current name of the item given
+mapWithName ::
+  Monad m =>
+  T tm ty sum ->
+  (Definition tm ty sum -> NameSpace.From Symbol -> T tm ty sum -> m (Definition tm ty sum)) ->
+  m (T tm ty sum)
+mapWithName t f = overTopLevelMap t (mapWithCurrentName f)
 
--- | @mapCurrentContext@ maps f over the entire context, this function
--- calls the function handed to it over the context. The function runs
--- over the current form and the context, so any dependencies may be
--- resolved and proper stored away.
--- mapCurrentContext' ::
--- Monad m => ContextForms m term ty sumRep -> T term ty sumRep -> m (T term ty sumRep)
-mapCurrentContext' ::
-  (Monad m, Show term, Show ty) =>
-  ( Maybe (Def term ty) ->
-    Symbol ->
-    Symbol ->
-    SumT term ty ->
-    T term ty sumRep ->
-    m (Maybe (Def term ty))
-  ) ->
+-- | @mapSumWithName@ maps a function over the sum constructors of the context
+mapSumWithName ::
+  Monad m =>
   T term ty sumRep ->
+  (SumT term ty -> Symbol -> T term ty sumRep -> m (Definition term ty sumRep)) ->
   m (T term ty sumRep)
-mapCurrentContext' f ctx@T {currentNameSpace, currentName} =
-  foldM dispatch ctx names
+mapSumWithName ctx f =
+  mapWithName ctx fOnSum
   where
-    names =
-      NameSpace.toList1FSymb (currentNameSpace ^. contents)
-    dispatch ctx (name, form) = do
-      case form of
-        Def d -> pure ctx
-        Record r -> do
-          -- Do the signature call in the module above and re-insert it
-          -- into the current module
-          let newTy = recordMTy r
-          -- now we should siwtch modules, note we are going to do a
-          -- direct insertion, so no need to reconstruct
-          let Just newCtx = inNameSpace (pure (NameSpace.extractValue name)) ctx
-          ctx' <- mapCurrentContext' f newCtx
-          let ctx'' = set (_currentNameSpace . mTy) newTy ctx'
-          -- just switch back to where we need to be
-          let Just finalCtx = inNameSpace (addTopName currentName) ctx''
-          pure finalCtx
-        Unknown u -> pure ctx
-        SumCon s@(Sum def name') -> do
-          newDef <- f def name' (NameSpace.extractValue name) s ctx
-          pure $ add name (SumCon (Sum newDef name')) ctx
-        TypeDeclar t -> pure ctx
-        CurrentNameSpace -> pure ctx
-        Information _ -> pure ctx
+    fOnSum (SumCon s) name ctx =
+      f s (NameSpace.extractValue name) ctx
+    fOnSum def _ _ = pure def
```

Comment:
Not handling Record?

Reply:
The record gets handled by the switcher, namely in mapWithCurrentName

```lisp
      -- tuple return is so we don't have to repeat work in recurseOnName
      (updatedCtx, defMaybeRecord) <- do
        newDef <- f form name ctx
        pure (add name newDef ctx, newDef)
      if
          | isRecord defMaybeRecord -> recurseOnName updatedCtx
          | otherwise -> pure updatedCtx
```
Now we don't have to worry about manually recursing it, we have a bit cleaner interface for that!